Tribler/community/multichain/block.py

```diff
@@ -0,0 +1,282 @@
+from hashlib import sha256
+from struct import pack_into, unpack_from, calcsize
+
+from Tribler.dispersy.crypto import ECCrypto
+
+# TODO: derive these at run time?
+HASH_LENGTH = 32
+SIG_LENGTH = 64
+PK_LENGTH = 74
+
+GENESIS_ID = '0'*HASH_LENGTH    # ID of the first block of the chain.
+GENESIS_SEQ = 1
+EMPTY_SIG = '0'*SIG_LENGTH
+EMPTY_PK = '0'*PK_LENGTH
+
+block_pack_format = "! I I Q Q {0}s I {0}s I {1}s {2}s".format(PK_LENGTH, HASH_LENGTH, SIG_LENGTH)
+block_pack_size = calcsize(block_pack_format)
+
+
+class MultiChainBlock(object):
+    """
+    Container for MultiChain block information
+    """
+
+    def __init__(self, data=None):
+        super(MultiChainBlock, self).__init__()
+        if data is None:
+            # data
+            self.up = self.down = 0
+            self.total_up = self.total_down = 0
+            # identity
+            self.public_key = EMPTY_PK
+            self.sequence_number = GENESIS_SEQ
+            # linked identity
+            self.link_public_key = EMPTY_PK
+            self.link_sequence_number = 0
+            # validation
+            self.previous_hash = GENESIS_ID
+            self.signature = EMPTY_SIG
+            # debug stuff
+            self.insert_time = None
+        else:
+            (self.up, self.down, self.total_up, self.total_down, self.public_key, self.sequence_number,
+             self.link_public_key, self.link_sequence_number, self.previous_hash, self.signature,
+             self.insert_time) = (data[0], data[1], data[2], data[3], data[4], data[5], data[6], data[7], data[8],
+                                  data[9], data[10])
+            if isinstance(self.public_key, buffer):
+                self.public_key = str(self.public_key)
+            if isinstance(self.link_public_key, buffer):
+                self.link_public_key = str(self.link_public_key)
+            if isinstance(self.previous_hash, buffer):
+                self.previous_hash = str(self.previous_hash)
+            if isinstance(self.signature, buffer):
+                self.signature = str(self.signature)
+
+    @property
+    def hash(self):
+        return sha256(self.pack()).digest()
+
+    def validate(self, database):
+        """
+        Validates this block against what is known in the database
+        :param database: the database to check against
+        :return: "valid" if the block does not violate any rules,
+                 "partial-next" if the block does not violate any rules, but there are missing blocks in the future
+                 "partial-prev" if the block does not violate any rules, but there are missing blocks in the past
+                 "partial" if the block does not violate any rules, but there are missing blocks on either side.
+                 "no-data" if there is not enough information known about the block to validate
+                 "invalid" if the block violate any of the rules,
```

Comment:
Maybe extract these validation results to constants, since we check for them elsewhere as well?